frontend/ui/data/onetime_migration.lua

```diff
@@ -0,0 +1,190 @@
+    if G_reader_settings:readSetting("screensaver_type") == "message" then
+        G_reader_settings:saveSetting("screensaver_type", "disable")
+        G_reader_settings:makeTrue("screensaver_show_message")
+    end
+    if G_reader_settings:has("screensaver_no_background") then
+        if G_reader_settings:isTrue("screensaver_no_background") then
+            G_reader_settings:saveSetting("screensaver_background", "none")
+        end
+        G_reader_settings:delSetting("screensaver_no_background")
+    end
+    if G_reader_settings:has("screensaver_white_background") then
+        if G_reader_settings:isTrue("screensaver_white_background") then
+            G_reader_settings:saveSetting("screensaver_background", "white")
+        end
+        G_reader_settings:delSetting("screensaver_white_background")
+    end
+end
+
+-- OPDS, same as above
+if from_version < Version:getNormalizedVersion("v2021.03") then
+    logger.info("Running one-time migration for v2021.03")
+
+    local opds_servers = G_reader_settings:readSetting("opds_servers")
+    if not opds_servers then
+        return
+    end
+
+    -- Update deprecated URLs & remove deprecated entries
+    for i = #opds_servers, 1, -1 do
+        local server = opds_servers[i]
+
+        if server.url == "http://bookserver.archive.org/catalog/" then
+            server.url = "https://bookserver.archive.org"
+        elseif server.url == "http://m.gutenberg.org/ebooks.opds/?format=opds" then
+            server.url = "https://m.gutenberg.org/ebooks.opds/?format=opds"
+        elseif server.url == "http://www.feedbooks.com/publicdomain/catalog.atom" then
+            server.url = "https://catalog.feedbooks.com/catalog/public_domain.atom"
+        end
+
+        if server.title == "Gallica [Fr] [Searchable]" or server.title == "Project Gutenberg [Searchable]" then
+            table.remove(opds_servers, i)
+        end
+    end
+    G_reader_settings:saveSetting("opds_servers", opds_servers)
+end
+
+-- Statistics, https://github.com/koreader/koreader/pull/7471
+if from_version < Version:getNormalizedVersion("v2021.03-12") then
+    logger.info("Running one-time migration for v2021.03-12")
+
+    local package_path = package.path
+    package.path = string.format("%s/?.lua;%s", "plugins/statistics.koplugin", package_path)
+    local ReaderStatistics = dofile("plugins/statistics.koplugin/main.lua")
+    package.path = package_path
+
+    local settings = G_reader_settings:readSetting("statistics", ReaderStatistics.default_settings)
+    -- Handle a snafu in 2021.03 that could lead to an empty settings table on fresh installs.
+    for k, v in pairs(ReaderStatistics.default_settings) do
+        if settings[k] == nil then
+            settings[k] = v
+        end
+    end
+    G_reader_settings:saveSetting("statistics", settings)
+end
+
+-- ScreenSaver, https://github.com/koreader/koreader/pull/7496
+if from_version < Version:getNormalizedVersion("v2021.03-35") then
```

Comment:
When making a PR that will need this kind of migration, might be tedious to guess the version it will start on (you'll have to guess what it would be, and make sure to not merge other PRs before yours).
Some other idea on how we will proceed?

Reply:
That's actually easy: take a [rancom commit](https://github.com/koreader/koreader/commit/2f9db25969f4af2f2961d45d981f57174ee638c0), then look at the bottom of the header:

![Screenshot_20210412_163222](https://user-images.githubusercontent.com/111974/114411793-cbc07500-9bac-11eb-960e-595c9f40d62d.png)

You've got a link to the branch, a link to the PR this commit is from... and a link to the first tag that implemented this commit ;).

But that's only if there *is* a "future" tag.

For "current" code, you can just take the revision of the last nightly and just append 1 ;).